server/channels/app/post.go

```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
 	}
 
 	permalinkPreviewedPost := post.GetPreviewPost()
-	for _, cm := range channelMembers {
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
-
-		postForUser := a.sanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedPost, permalinkPreviewedChannel, cm.UserId)
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	} else {
+		return false, nil
+	}
 
-		// Using DeepCopy here to avoid a race condition
-		// between publishing the event and setting the "post" data value below.
-		messageCopy := message.DeepCopy()
-		broadcastCopy := messageCopy.GetBroadcast()
-		broadcastCopy.UserId = cm.UserId
-		messageCopy.SetBroadcast(broadcastCopy)
+	postCopy := post.Clone()
 
-		postJSON, jsonErr := postForUser.ToJSON()
-		if jsonErr != nil {
-			mlog.Warn("Failed to encode post to JSON", mlog.Err(jsonErr))
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
+	}
```

Comment:
> Then posting this shouldn't expand that post for them.

Oh, got it: so the caller of `createPost` is getting this post object back, and since we've set `post.Metadata.Embeds[0].Data` above, we need to unset it here after making a copy. But that means in the happy case when the user does have access, this method also has the surprising side effect of resolving `post.Metadata.Embeds[0].Data` via a method called `publishWebsocketEventForPermalinkPost`.

Looking at the old code, it's not obvious "where" this used to happen. Does it stand out to you?

Reply:
It used to happen in `sanitizePostMetadataForUserAndChannel`. I agree that this has become slightly confusing. I'll look into this separately in a further PR. Thanks for raising this.